optaplanner-test/src/main/java/org/optaplanner/test/impl/score/stream/SingleConstraintAssertion.java

```diff
@@ -0,0 +1,223 @@
+                .sum();
+    }
+
+    private static void assertCorrectMatchWeight(Number matchWeightTotal) {
+        if (matchWeightTotal.doubleValue() <= 0) {
+            throw new IllegalArgumentException("Expected a positive match weight, given (" + matchWeightTotal + ").");
+        }
+    }
+
+    private void assertImpact(ScoreImpactType scoreImpactType, Number weight, String message) {
+        Number impact = getImpact();
+        AbstractConstraint<?, ?> constraint = (AbstractConstraint<?, ?>) getParentConstraintVerifier().getConstraint();
+        // Null means we're just looking for any kind of penalty or an impact.
+        boolean isCorrectImpactType = scoreImpactType == null || scoreImpactType == constraint.getScoreImpactType();
+        if (isCorrectImpactType && weight.equals(impact)) {
+            return;
+        }
+        String constraintId = constraint.getConstraintId();
+        String assertionMessage = getAssertionErrorMessage(scoreImpactType, weight, constraint.getScoreImpactType(),
+                impact, constraintId, message);
+        throw new AssertionError(assertionMessage);
+    }
+
+    private static String getAssertionErrorMessage(ScoreImpactType expectedImpactType, Number expectedImpact,
+            ScoreImpactType actualImpactType, Number actualImpact, String constraintId, String message) {
+        boolean hasMessage = message != null;
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream(); PrintStream printStream = new PrintStream(baos)) {
+            String preformattedMessage = "Broken expectation.%n" +
+                    "%18s: %s%n" +
+                    (hasMessage ? "%18s: %s%n" : "") +
+                    "%18s: %s (%s)%n" +
+                    "%18s: %s (%s)";
+            String expectedImpactLabel = "Expected " + getImpactTypeLabel(expectedImpactType);
+            String actualImpactLabel = "Actual " + getImpactTypeLabel(actualImpactType);
+            if (hasMessage) {
+                printStream.printf(preformattedMessage,
+                        "Message", message,
+                        "Constraint", constraintId,
+                        expectedImpactLabel, expectedImpact, expectedImpact.getClass(),
+                        actualImpactLabel, actualImpact, actualImpact.getClass());
+
+            } else {
+                printStream.printf(preformattedMessage,
+                        "Constraint", constraintId,
+                        expectedImpactLabel, expectedImpact, expectedImpact.getClass(),
+                        actualImpactLabel, actualImpact, actualImpact.getClass());
+            }
+            return baos.toString();
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed assembling asserting message.", e);
+        }
+    }
+
+    private static String getImpactTypeLabel(ScoreImpactType scoreImpactType) {
+        switch (scoreImpactType) {
```

Comment:
Throws NPE if you `expectNoImpact()` but the constraint does a penalty or a reward. To reproduce, change any `expectPenalty()` in the NQueens test to `expectNoImpact()`.

```
java.lang.IllegalStateException: Failed assembling asserting message.

	at org.optaplanner.test.impl.score.stream.SingleConstraintAssertion.getAssertionErrorMessage(SingleConstraintAssertion.java:93)
	at org.optaplanner.test.impl.score.stream.SingleConstraintAssertion.assertImpact(SingleConstraintAssertion.java:62)
	at org.optaplanner.test.impl.score.stream.SingleConstraintAssertion.expectNoImpact(SingleConstraintAssertion.java:213)
	at org.optaplanner.test.impl.score.stream.SingleConstraintAssertion.expectNoImpact(SingleConstraintAssertion.java:220)
	at org.optaplanner.examples.nqueens.solver.score.NQueensConstraintProviderTest.horizontalConflictWithTwoQueens(NQueensConstraintProviderTest.java:62)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)
Caused by: java.lang.NullPointerException
	at org.optaplanner.test.impl.score.stream.SingleConstraintAssertion.getImpactTypeLabel(SingleConstraintAssertion.java:98)
	at org.optaplanner.test.impl.score.stream.SingleConstraintAssertion.getAssertionErrorMessage(SingleConstraintAssertion.java:76)
	... 31 more
```